python/scannerpy/database.py

```diff
@@ -244,14 +269,87 @@ def start_cluster(self, master, workers):
             master: ssh-able address of the master node.
             workers: list of ssh-able addresses of the worker nodes.
         """
-        master_cmd = 'python -c "from scannerpy import Database as Db; Db().start_master()"'
-        worker_cmd = 'python -c "from scannerpy import Database as Db; Db().start_worker()"'
-
-        master = self._run_remote_cmd(master, master_cmd)
-        workers = [self._run_remote_cmd(w, worker_cmd) for w in workers]
-        master.wait()
-        for worker in workers:
-            worker.wait()
+
+        if master is None:
+            self._master_address = self.config.master_address
+        else:
+            self._master_address = master
+        if workers is None:
+            self._worker_addresses = [self.config.worker_address]
+        else:
+            self._worker_addresses = workers
+
+        master_cmd = ('python -c ' +
+                      '"from scannerpy import start_master\n' +
+                      'start_master(block=True)"')
+        worker_cmd = ('python -c ' +
+                      '"from scannerpy import start_worker\n' +
+                      'start_worker(\'{:s}\', block=True)"').format(
+                          self._master_address)
+
+        self._master_conn = self._run_remote_cmd(self._master_address,
+                                                 master_cmd)
+        # Wait for master to start
+        slept_so_far = 0
+        sleep_time = 20
+        while slept_so_far < sleep_time:
+            if self._connect_to_master():
+                break
+            time.sleep(0.3)
+            slept_so_far += 0.3
```

Comment:
Probably better to check the wall clock time (`now()`) instead of assuming it sleeps exactly as long as you asked for, although I doubt it would be a big difference.